node/src/components/network.rs

```diff
@@ -0,0 +1,610 @@
+    }
+
+    fn handle_connection_established(
+        &mut self,
+        peer_id: NodeId,
+        endpoint: ConnectedPoint,
+        num_established: NonZeroU32,
+    ) -> Effects<Event<P>> {
+        debug!(%peer_id, ?endpoint, %num_established,"{}: connection established", self.our_id);
+        let _ = self.peers.insert(peer_id);
+
+        if let ConnectedPoint::Dialer { address } = endpoint {
+            if let Some(state) = self.known_addresses.get_mut(&address) {
+                if *state == ConnectionState::Pending {
+                    *state = ConnectionState::Connected
+                }
+            }
+        }
+
+        Effects::new()
+    }
+
+    fn handle_connection_closed(
+        &mut self,
+        peer_id: NodeId,
+        endpoint: ConnectedPoint,
+        num_established: u32,
+        cause: Option<String>,
+    ) -> Effects<Event<P>> {
+        if num_established == 0 {
+            let _ = self.peers.remove(&peer_id);
+        }
+        debug!(%peer_id, ?endpoint, %num_established, ?cause, "{}: connection closed", self.our_id);
+        Effects::new()
+    }
+
+    fn handle_unreachable_address(
+        &mut self,
+        peer_id: NodeId,
+        address: Multiaddr,
+        error: PendingConnectionError<io::Error>,
+        attempts_remaining: u32,
+    ) -> Effects<Event<P>> {
+        debug!(%peer_id, %address, %error, %attempts_remaining, "{}: failed to connect", self.our_id);
+        Effects::new()
+    }
+
+    fn handle_unknown_peer_unreachable_address(
+        &mut self,
+        effect_builder: EffectBuilder<REv>,
+        address: Multiaddr,
+        error: PendingConnectionError<io::Error>,
+    ) -> Effects<Event<P>> {
+        debug!(%address, %error, "{}: failed to connect", self.our_id);
+        if let Some(state) = self.known_addresses.get_mut(&address) {
+            if *state == ConnectionState::Pending {
+                *state = ConnectionState::Failed
+            }
+        }
+
+        // TODO - Consider how to handle becoming isolated _after_ we've bootstrapped.  Ensure we
+        //        can't re-enter this block unless we implement re-bootstrapping for example.
+        if self.is_isolated() {
+            if self.is_bootstrap_node() {
+                info!(
+                    "{}: failed to bootstrap to any other nodes, but continuing to run as we are a \
+                    bootstrap node", self.our_id
+                );
+            } else {
+                // Note that we could retry the connection to other nodes, but for now we just
+                // leave it up to the node operator to restart.
+                return fatal!(
+                    effect_builder,
+                    "{}: failed to connect to any known node, now isolated",
+                    self.our_id
+                );
+            }
+        }
+        Effects::new()
+    }
+
+    fn handle_new_listen_address(&mut self, address: Multiaddr) -> Effects<Event<P>> {
```

Comment:
I think these functions could easily be written inline in the `match` statement, to cut down on clutter.

Reply:
I have a preference for keeping as much logic out of the `handle_event` as possible, delegating to helpers instead.  Again, this is a fairly common pattern in our codebase, and decreases clutter in my opinion (and also in Ed's opinion I believe).

Reply:
I'm kind in the opposite camp. If it is < 8-10 lines, much better to have it inline, as the overhead of adding a function is typically larger than that.

Reply:
OK - I'll make the change.